Add SlackLogger w/ custom formatting to composer/logger

waiwuc opened the pull request:
# What does this PR do?
1. Allows training metrics to be output to a slack channel using [slack webhooks](https://api.slack.com/messaging/webhooks. Create your own slack app and add an incoming webhook url, then export the url as an envar `SLACK_WEBHOOK_URL`. 
2. Allows slack message to be formatted using slack [block kit builder ](https://app.slack.com/block-kit-builder/T01G9MS9XUG). Pass in a lambda function that formats your metrics according to block kit builder guidelines. 
3. If formatter function is not passed in, default to logging a header + text section where keys are bolded and values are unformatted. 
4. Since formatter function can be passed into `SlackLogger` as a first class function, more complex layouts can be created easily. 

NOTE: See example use case in `example/slack_log_metrics.py`. 

## Examples 
- With simple formatting (just text section where keys are bolded) 
<img width="283" alt="Screenshot 2023-04-10 at 4 13 54 PM" src="https://user-images.githubusercontent.com/92327646/231019028-5abe2532-bda7-446b-9653-44ca15a4a103.png">
- With header section (steps as headers) 
<img width="212" alt="Screenshot 2023-04-10 at 4 14 17 PM" src="https://user-images.githubusercontent.com/92327646/231019065-2fd27059-2912-4e44-91a4-f1ce39634a48.png">
<img width="221" alt="Screenshot 2023-04-10 at 4 14 22 PM" src="https://user-images.githubusercontent.com/92327646/231019075-d54de2de-c10b-48f8-a74a-0f594ae46bb6.png">

dakinggg commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,67 @@
+from composer.loggers.logger_destination import LoggerDestination
+
+
+class SlackLogger(LoggerDestination):
```
Should we add some `filter_keys` option or something? Otherwise this is going to log _everything_ to slack right?

waiwuc replied:
done!

dakinggg commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,67 @@
+        return blocks
+        
+    def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None) -> None:
+        self.logged_metrics.update(metrics)
```
what is `self.logged_metrics` used for?

waiwuc replied:
excess, i just removed it

dakinggg commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,67 @@
+
+    # Rich message layouts can be created using Slack Blocks Kit.
+    # See documentation here: https://api.slack.com/messaging/composing/layouts
+    def _log_to_slack(
```
I think this function should either not be on the class, and just accept all the args (i.e. not use `self.webhook_url`), or use everything via `self` (i.e. `self.client`, `self.log_metrics_formatter_func`, `self.webhook_url`).

waiwuc replied:
I updated to use this function for metrics, traces and hparams, so I needed to parametrize it

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,67 @@
+        webhook_url: str,
+        log_metrics_formatter_func: Optional[Callable] = None,
+    ) -> None:
+        self.webhook_url = webhook_url
```
What do you think of webhook url being optional if the user sets the SLACK_WEBHOOK_URL env variable. Then it would be a lot easier for someone to write a script because they wouldn't have to go look for it every time or save it to their script

eracah commented on examples/slack_log_metrics.py:
```diff
@@ -0,0 +1,41 @@
+train_dataloader = DataLoader(dataset, batch_size=128)
+
+SLACK_WEBHOOK_URL=os.getenv('SLACK_WEBHOOK_URL')
+
```
should we also include an example for using the slack logger with a callback?

dakinggg commented on composer/loggers/__init__.py:
```diff
@@ -21,22 +21,12 @@
-    'MLFlowLogger',
-    'SlackLogger'
+    'Logger', 'LoggerDestination', 'FileLogger', 'InMemoryLogger', 'ProgressBarLogger', 'WandBLogger',
+    'RemoteUploaderDownloader', 'TensorboardLogger', 'CometMLLogger', 'ConsoleLogger', 'MLFlowLogger', 'SlackLogger'
```
please add a trailing comma back here to keep the formatting

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,64 @@
+    def __init__(
+        self,
+        webhook_url: str = '',
+        log_metrics_formatter_func: Optional[Callable] = None,
```
should we add a log_interval here, where the user can specify how often to log a metric to slack

waiwuc replied:
let me do that, I think I could add it in the formatter function as well. Ideally have one param to be one configuration with both formatting + interval

waiwuc replied:
I will defer this to another iteration, interval is a little harder to code up. Instead ill have an include_keys config that allows us to log only essential keys, so we minimize the log spam

eracah reviewed (commented):
Looks pretty good. Thanks, Wai. Just a few a's.

waiwuc commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,64 @@
+    def __init__(
+        self,
+        webhook_url: str = '',
+        log_metrics_formatter_func: Optional[Callable] = None,
```
`log_metrics_formatter_func` is actually variadic, since we pass in `data` as the only required positional argument and `**kwargs`

There should be a way to cast this Callable as variadic though, let me look into it

waiwuc commented on composer/loggers/slack_logger.py:
```diff
@@ -1,64 +1,192 @@
+        except ImportError as e:
+            raise MissingConditionalImportError('health_checker', 'slack_sdk', None) from e
+
+        self.client = WebClient()
+        self.slack_logging_api_key = os.environ.get('SLACK_LOGGING_API_KEY', None)
+        self.channel_id = os.environ.get('SLACK_LOGGING_CHANNEL_ID', None)
```
Changed this to use the post message api instead of webhooks, because this allows user to send messages to a channel directly.

Webhooks with custom channels has been deprecated, see https://glitch.com/~slack-channel-webhook-blueprint

waiwuc commented on composer/loggers/slack_logger.py:
```diff
@@ -1,64 +1,192 @@
+            raise RuntimeError(
+                'SLACK_LOGGING_API_KEY and SLACK_LOGGING_CHANNEL_ID must be set as environment variables')
+
+        self.log_metrics_config = log_metrics_config
+        self.log_traces_config = log_traces_config
+        self.log_hparams_config = log_hparams_config
+
+        # Maintain logger state
+        now = time.time()
+        self.logs_since_last_log, self.last_log_time = list(Any), now
+        self.traces_since_last_log, self.last_trace_time = list(Any), now
+        self.hparams_since_last_log, self.last_hparams_time = list(Any), now
```
Allows different time intervals and formats for sending logs / traces / hparams. In most cases, user would implement just one out of these 3

waiwuc commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,67 @@
+         
+    Args:
+        webhook_url (str): Slack Webhook URL
+        log_metrics_formatter_func ((...) -> Any | None): A formatter function that returns list of blocks to be sent to slack
```
Pulled up to doc string

waiwuc commented on examples/slack_log_metrics.py:
```diff
@@ -0,0 +1,41 @@
+from torchvision import datasets, transforms
```
done

waiwuc commented:
@eracah , @dakinggg , thanks for the reviews! 

This should be ready for another review - the major things changed are
1. Use Slack's `postMessage` api instead of webhooks. This allows user to send messages to a channel directly. Webhooks with custom channels has been deprecated, see https://glitch.com/~slack-channel-webhook-blueprint
2. implement `log_traces` and `log_hparam` in addition to `log_metrics`. This design is now generic enough to log all 3 types with different configs. 
3. Update formatter functions to configs that consist of intervals, format function, and filter keys, which allow users more control over which keys to log in what format and how frequently. 

Appreciate your reviews a ton :)

dakinggg commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,190 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Log metrics to slack, using Slack postMessage api - https://api.slack.com/methods/chat.postMessage.
```
Could you please move all of this into the `SlackLogger` class docstring? I'm pretty sure it needs to be there for our docs to pick it up. (although if you build the docs locally and it shows up, that is also fine)

waiwuc replied:
moved

dakinggg commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,190 @@
+    """Log config for slack logger, including formatter function, keys to include, and interval in seconds.
+
+    formatter_func ((...) -> Any | None): A formatter function that returns list of blocks to be sent to slack
+    include_keys (Set[str]): A set of metric/logs/traces keys to include in the message. If None, no keys are included.
```
what do you think about defaulting to all keys if `include_keys=None`. I think that behavior might be more useful as a default? not completely sure though

waiwuc replied:
yeah good point, None and empty should probably have different behavior, even though there's no use case for an empty set. 

Will make the change and update the docs.

dakinggg commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,190 @@
+    formatter_func ((...) -> Any | None): A formatter function that returns list of blocks to be sent to slack
+    include_keys (Set[str]): A set of metric/logs/traces keys to include in the message. If None, no keys are included.
+    interval_in_seconds (int): Interval in seconds to send logs to slack. If None, message is sent after every log call.
+    Note that no more than 50 items are allowed to send in a single message.
```
Should we flush after 50 items are present regardless of time interval?

waiwuc replied:
yes that would be better behavior than crashing the run, let me add this change

dakinggg commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,190 @@
+        filtered_data = {k: v for k, v in data.items() if k in config.include_keys} if config.include_keys else data
+        blocks = config.formatter_func(
+            filtered_data, **
+            kwargs) if config.formatter_func else self._default_log_bold_key_normal_value_pair_with_header(
```
```suggestion
            kwargs) if config.formatter_func is not None else self._default_log_bold_key_normal_value_pair_with_header(
```

dakinggg commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,190 @@
+        if len(updated_blocks_to_log) > 50:
+            raise RuntimeError('Cannot send more than 50 items in a single slack message')
+
+        if len(updated_blocks_to_log) > 0 and self.channel_id and (
```
```suggestion
        if len(updated_blocks_to_log) > 0 and self.channel_id is not None and (
```

dakinggg reviewed (approved):
LGTM! I would like some unit tests (with the slack client mocked), but am also ok if you'd prefer to get it in so people can start playing with it.

dakinggg commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,190 @@
+
+        updated_blocks_to_log = blocks + blocks_to_log
+
+        if len(updated_blocks_to_log) > 50:
```
should we send in groups of 50 rather than just erroring out? or at least truncate and warn rather than error. I would not be super happy if the slack logger taking in too many messages crashed my run :P

waiwuc replied:
Ill make the change to flush the logs when exceeding 50. 

Was probably also a mistake to pick `RuntimeError` for these types of errors. Havent looked into the error boundaries in trainer, but wondering if we've thought of having a `LoggerError` class that could be caught and printed to std out rather than throwing?

dakinggg replied:
we could probably be better with our errors. I think we mostly just abuse the standard errors and don't do much special try/catch like that

waiwuc commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,190 @@
+        self.channel_id = os.environ.get('SLACK_LOGGING_CHANNEL_ID', None)
+
+        if self.slack_logging_api_key is None or self.channel_id is None:
+            raise RuntimeError(
```
good point

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,190 @@
+
+    def __init__(
+        self,
+        log_metrics_config: Optional[SlackLogConfig] = None,
+        log_traces_config: Optional[SlackLogConfig] = None,
+        log_hparams_config: Optional[SlackLogConfig] = None,
```
```suggestion
        formatter_func: Optional[Callable[..., List[Dict[str, Any]]]] = None,,
        include_keys: Optional[Set[str]] = None,
        interval_in_seconds: Optional[int] = None,
```

I think having the SlackLogConfig for each category logged is a bit of overkill:

1. It is kinda an unnecessary lift for a user learn about this new class and instantiate a new object for every type of filtering they want to do.
2. It is probably overkill to have a different config for traces, hyperparameters, and metrics. (Traces are only really used as debugging and it is easy for a user to mistake certain things as hyperparameters, which are actually metrics, so the user would be playing a game of guess and check to set the right config.
3. Seems like we want just an MVP. This seems like a premature abstraction

We could also a do a dictionary instead of flat arguments

waiwuc replied:
yes that makes sense, updated!

eracah reviewed (changes requested):
Looks pretty good, but I really think we should simplify the logging config part. It would be simpler and cover almost every situation just to have one set of `formatter_func`, `include_keys`, and `interval_in_seconds`. 

Also this would set an unfair precedent for other loggers to use this config structure, which I think would overcomplicate and boiler-platify the rest of the Logger codebase

waiwuc commented:
removed configs for hparams/metrics/traces, we can configure the `SlackLogger` directly, and use one config for all 3 types of logs. 

Also added a bunch of unit tests

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,143 @@
+        self.slack_logging_api_key = os.environ.get('SLACK_LOGGING_API_KEY', None)
+        self.channel_id = os.environ.get('SLACK_LOGGING_CHANNEL_ID', None)
+
+        if self.slack_logging_api_key is None:
+            raise RuntimeError('SLACK_LOGGING_API_KEY must be set as environment variable')
+        if self.channel_id is None:
+            raise RuntimeError('SLACK_LOGGING_CHANNEL_ID must be set as environment variable')
```
We could do something where we allow the user to optionally pass these into the constructor and if they are not passed in then we look at the env variables. Wdyt?

waiwuc replied:
I can make that change, but not convinced about the use case for having these as parameters in the init function. These slack keys should be treated as secrets and if so, they are best set as env variables.

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,143 @@
+
+    Args:
+        formatter_func ((...) -> Any | None): A formatter function that returns list of blocks to be sent to slack.
+        include_keys (Set[str]): A set of metric/logs/traces keys to include in the message. If None, all keys are included.
```
I think we might want to tweak the defaults here. I don't think by default all keys should be included. It might be too easy for a user to just pollute their slack channel accidentally. We might want to make this a required argument. If the user passes in a sequence, then those are the keys logged and if they pass in 'all' or '*' or something then all keys are logged

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,143 @@
+    def __init__(
+        self,
+        formatter_func: Optional[Callable[..., List[Dict[str, Any]]]] = None,
+        include_keys: Optional[Set[str]] = None,
```
I don't think this should be a set. Let's make it a `Sequence` and then we can cast it to set a inside the function

waiwuc replied:
Ill make the update, but curious why not a `Set`?

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,143 @@
+            kwargs) if self.formatter_func is not None else self._default_log_bold_key_normal_value_pair_with_header(
+                filtered_data, **kwargs)
+
+        self.logs += blocks
```
hmmm I'm not sure appending to a list is the best approach here because if the user sets an infrequent interval, they will see pretty stale data and if they set a high max_logs_per_message they will get overwhelmed with a full history for every key. 
I think we may want to either have a queue or remove values from the list based on a separate dictionary of the keys that are "pending" to be logged and if a new value for that key comes in before that key, value pair gets posted, then we overwrite it?

waiwuc replied:
if they set a high `max_logs_per_message`, we will still flush the logs when it hits 50 as that's the maximum that we can send to slack client. 

I am conflicted about having a full history for every key. Conceptually, if we wanted to track something like loss spikes, would we not need to have a full history of total loss at every step? On the other hand, we have also decided against version history for run metadata (although maybe that's more because of technical limitation vs desired product behavior) - cc @mvpatel2000 for thoughts here too. 

Given slack is high noise channel, maybe we'd optimize for small number of keys but larger number of values logged over time, rather than large number of keys and snapshots being logged?

waiwuc replied:
Synced quickly with @eracah , we can think about adding `log_full_history: Optional[bool]=False` for future, just launch this first without version history.

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,143 @@
+        self.logs += blocks
+        now = time.time()
+
+        if len(self.logs) > 0 and self.channel_id is not None and (
+                len(self.logs) >= self.max_logs_per_message or self.interval_in_seconds is None or
+                now - self.last_log_time >= self.interval_in_seconds):
```
this is a pretty complicated boolean. can we maybe split it up into a few boolean variables?

I think `self.channel_id is not None` is unnecessary since the code errors out if it is None right?
```suggestion
        interval_has_elapsed = (self.interval_in_seconds is None or (now - self.last_log_time) >= self.interval_in_seconds)
        if len(self.logs) >= self.max_logs_per_message or (interval_has_elapsed and len(self.logs) > 0): 
```
or you could even set self.interval_in_seconds to 0 if it is passed in as None

waiwuc replied:
yes let me split up into a few variables. I needed to add `self.channel_id is not None` for typing reasons

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,143 @@
+            self.client.chat_postMessage(token=f'{self.slack_logging_api_key}',
+                                         channel=self.channel_id,
+                                         blocks=self.logs[:self.max_logs_per_message],
+                                         text=str(self.logs[:self.max_logs_per_message]))
```
the `text` variable is supposed to be the formatted text in either plaintext or markdown right? it's not supposed be just a stringification of the block right? https://api.slack.com/methods/chat.postMessage#blocks_and_attachments

waiwuc replied:
`When you're using blocks in your message payload, the top-level text field becomes a fallback message displayed in notifications. Blocks should define everything else about the desired visible message.` 

text is there for notifications actually. I didn't test this behavior manually, but I definitely should, and probably just updated to a shorter string like f"Value for {len(self.logs)} metrics updated"

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,143 @@
+        self,
+        formatter_func: Optional[Callable[..., List[Dict[str, Any]]]] = None,
+        include_keys: Optional[Set[str]] = None,
+        interval_in_seconds: Optional[int] = None,
```
Should this be seconds? That's not the most intuitive thing for the user. We may want to use the Time abstraction. See how it is done in ConsoleLogger for reference: https://github.com/mosaicml/composer/blob/dev/composer/loggers/console_logger.py#L42

waiwuc replied:
great call out, I like that better, let me update this

eracah reviewed (changes requested):
Hey Wai, looks pretty good! Thanks for addressing my feedback from before, but a couple more bits of feedback for you.
Also, can you post the code that output those example images (you can redact the API key of course)? Also, thanks for adding those unit tests!

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,219 @@
+            self._format_and_send_blocks_to_slack(max_log_entries_dict, **kwargs)
+            inx += self.max_logs_per_message
+
+        # max_log_entries_dict = dict(itertools.islice(self.log_dict.items(), 0, 1))
```
remove old comments

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,219 @@
+            raise TypeError
+
+        inx = 0
+        while inx < len(self.log_dict.keys()):
```
this just sends max_logs_per_message blocks at a time, but it still sends all the messages?

waiwuc replied:
if there's 1000 metrics in between batches, this will flush all 1000 metrics 50 at a time

eracah commented on tests/loggers/test_slack_logger.py:
```diff
@@ -0,0 +1,57 @@
+from tests.common import RandomClassificationDataset, SimpleModel
+
+
+@pytest.mark.parametrize('include_keys', [['loss*']])
+def test_slack_logger_metrics(include_keys: Set[str]):
```
```suggestion
@pytest.mark.parametrize('include_keys', [['loss*']])
@pytest.mark.parametrize('interval', ['1ba', '1ep'])
def test_slack_logger_metrics(include_keys: Set[str], interval: str):
```

eracah commented on tests/loggers/test_slack_logger.py:
```diff
@@ -0,0 +1,57 @@
+            }
+        } for k, v in data.items()]),
+        include_keys=include_keys,
+        log_interval='1ep',
```
```suggestion
        log_interval=interval,
```

eracah commented on tests/loggers/test_slack_logger.py:
```diff
@@ -0,0 +1,57 @@
+            model=SimpleModel(),
+            train_dataloader=DataLoader(RandomClassificationDataset()),
+            train_subset_num_batches=2,
+            max_duration='1ep',
```
```suggestion
            max_duration=interval,
```

eracah commented on tests/loggers/test_slack_logger.py:
```diff
@@ -0,0 +1,57 @@
+            loggers=[slack_logger],
+        )
+
+    # Log interval is 1 epoch, so no logs should be sent before fit
```
```suggestion
    # Log interval is 1 epoch or 1 batch, so no logs should be sent before fit
```

eracah commented on composer/loggers/slack_logger.py:
```diff
@@ -0,0 +1,216 @@
+        Args:
+            data (Dict[str, Any]): Data to be logged.
+            **kwargs: Additional arguments to be passed to the formatter function
+            (Only support "header" argument now)
```
```suggestion
            (Only supports "header" argument now)
```

eracah reviewed (approved):
LGTM! Thanks, @waiwuc great work!
